pkg/autodiscovery/providers/kubecontainer.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2016-present Datadog, Inc.
+
+//go:build !serverless
+// +build !serverless
+
+package providers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/common/utils"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/providers/names"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/containers"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/workloadmeta"
+)
+
+// KubeContainerConfigProvider implements the ConfigProvider interface for both kubelet and containers
+type KubeContainerConfigProvider struct {
+	workloadmetaStore workloadmeta.Store
+	podCache          map[string]*workloadmeta.KubernetesPod
+	containerCache    map[string]*workloadmeta.Container
+	configErrors      map[string]ErrorMsgSet
+	upToDate          bool
+	streaming         bool
+	once              sync.Once
+	sync.RWMutex
+}
+
+// NewKubeContainerConfigProvider returns a new ConfigProvider subscribed to both container
+// and pods
+func NewKubeContainerConfigProvider(*config.ConfigurationProviders) (ConfigProvider, error) {
+	return &KubeContainerConfigProvider{
+		workloadmetaStore: workloadmeta.GetGlobalStore(),
+		configErrors:      make(map[string]ErrorMsgSet),
+		podCache:          make(map[string]*workloadmeta.KubernetesPod),
+		containerCache:    make(map[string]*workloadmeta.Container),
+	}, nil
+}
+
+// String returns a string representation of the KubeContainerConfigProvider
+func (k *KubeContainerConfigProvider) String() string {
+	return names.KubeContainer
+}
+
+// Collect retrieves all running pods and extract AD templates from their annotations.
+func (k *KubeContainerConfigProvider) Collect(ctx context.Context) ([]integration.Config, error) {
+	k.once.Do(func() {
+		go k.listen()
+	})
+
+	k.Lock()
+	k.upToDate = true
+	k.Unlock()
+
+	return k.generateConfigs()
+}
+
+func (k *KubeContainerConfigProvider) listen() {
+	const name = "ad-kubecontainerprovider"
+
+	k.Lock()
+	k.streaming = true
+	health := health.RegisterLiveness(name)
+	defer func() {
+		err := health.Deregister()
+		if err != nil {
+			log.Warnf("error de-registering health check: %s", err)
+		}
+	}()
+	k.Unlock()
+
+	podCh := k.workloadmetaStore.Subscribe(name, workloadmeta.NormalPriority, workloadmeta.NewFilter(
+		[]workloadmeta.Kind{workloadmeta.KindKubernetesPod},
+		workloadmeta.SourceNodeOrchestrator,
+		workloadmeta.EventTypeAll,
+	))
+
+	contCh := k.workloadmetaStore.Subscribe(name, workloadmeta.NormalPriority, workloadmeta.NewFilter(
+		[]workloadmeta.Kind{workloadmeta.KindContainer},
+		workloadmeta.SourceRuntime,
+		workloadmeta.EventTypeAll,
+	))
```

Comment:
It's better to do a single subscription, that filters for `[]workloadmeta.Kind{workloadmeta.KindKubernetesPod, workloadmeta.KindContainer}` and `workloadmeta.SourceAll`. The end result is equivalent, there's no need to separate event processing between runtime and node orchestrator sources.

Reply:
That was my first draft actually (see [https://github.com/DataDog/datadog-agent/pull/12851/commits/c805a0711a3b34fbd340e03e8b1ca11e94978b36](https://github.com/DataDog/datadog-agent/pull/12851/commits/c805a0711a3b34fbd340e03e8b1ca11e94978b36)).

I somehow convinced myself that there could be a race where a container would be added from source `runtime` and then later added again from source `node_orchestrator`, where we may lose the data from the original `runtime` event.

After checking through this again I don't see how this is possible because with `SourceAll` we get the merged events from wlm, and assuming the merge operation is correct, we would never lose data. Even if two containers are added in separate notifications, the second should always be a superset of the first.

I'll revert this second commit that does separate subscriptions.